First draft updated guide on hooks (#376)
astronomer/airflow-guides: guides/what-is-a-hook.md

---
title: "Hooks 101"
description: "An introduction to Hooks in Apache Airflow."

> date: 2022-05-24T00:00:00.000Z

Author: That does make sense, I changed it back :)

Hooks were used in this case because none of the existing S3 Operators fit the use case of reading data from several files within an S3 bucket. Similarly none of the existing Slack operators are able to return the response of a Slack API call, which you may want to log for monitoring purposes.

The full source code of the Hooks used can be found here: [S3Hook source code](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/s3.py), [SlackHook source code](https://github.com/apache/airflow/blob/main/airflow/providers/slack/hooks/slack.py).

Start by making sure you have the necessary Airflow Providers installed. Either by running:

```console
pip install apache-airflow-providers-amazon
pip install apache-airflow-providers-slack
```

or, if you are using the Astro CLI by adding the following packages in requirements.txt:

```text
apache-airflow-providers-amazon
apache-airflow-providers-slack
```

> Next you will need to set up the connection to the S3 bucket and Slack in the Airflow UI (localhost:8080 by default).

Reviewer: ```suggestion
Next you will need to set up the connection to the S3 bucket and Slack in the Airflow UI.
```

# and returning the server response
@task.python
def post_to_slack(sum_check_result):
    slack_hook = SlackHook(slack_conn_id='hook_tutorial_slack_conn')

    if sum_check_result[0] == True:
        server_response = slack_hook.call(api_method='chat.postMessage',
                        json={"channel": "#test-airflow",
                        "text": f"""All is well in your bucket!
                        Correct sum: {sum_check_result[1]}!"""})
    else:
        server_response = slack_hook.call(api_method='chat.postMessage',
                        json={"channel": "#test-airflow",
                        "text": f"""A test on your bucket contents failed!
                        Target sum not reached: {sum_check_result[1]}"""})

    # return the response of the API call (for logging or use downstream)
    return server_response

# implementing the DAG
with DAG(dag_id='hook_tutorial',
        start_date=datetime(2022,5,20),
        schedule_interval='@daily',
        catchup=False,
        ) as dag:

    # the dependencies are automatically set by XCom
    response = read_keys_form_s3()
    sum_check_result = run_sum_check(response)
    post_to_slack(sum_check_result)

> ```

Reviewer: Maybe just add one final sentence or two here describing the specific Hooks used for accomplishing the use case? Check out this guide as an example (it describes the DAG's steps before the code, which would also be fine): https://www.astronomer.io/guides/dynamic-tasks#example-implementations

Author: I added a description of the DAGs steps at the end of the document and included which Hooks were used with which methods there.

To run this example DAG, start by making sure you have the necessary Airflow Providers installed. If you are using the Astro CLI, you can add the following packages to your `requirements.txt`:

```text
apache-airflow-providers-amazon
apache-airflow-providers-slack
```

Next you will need to set up the connection to the S3 bucket and Slack in the Airflow UI.

1. Navigate to Admin -> Connections and click on the plus sign to add a new connection.
2. Select 'Amazon S3' as connection type for the S3 bucket (if the connection type is not showing up, double check that you installed the provider correctly) and provide the connection with your AWS Access key ID as login and your AWS Secret access key as password ([How to get your AWS access key ID and secret access key](https://docs.aws.amazon.com/powershell/latest/userguide/pstools-appendix-sign-up.html)).
3. For the connection to Slack select 'Slack Webhook' as the connection type and provide your [Bot User OAuth Token](https://api.slack.com/authentication/oauth-v2) as a password. This token can be obtained by navigating to the 'OAuth & Permissions tab' under 'Features' on api.slack.com/apps.

> The DAG below uses [Airflow Decorators](https://registry.astronomer.io/guides/airflow-decorators) to define tasks and [XCom](https://registry.astronomer.io/guides/airflow-passing-data-between-tasks) to pass information between them. The name of the S3 bucket and the names of the files that the first task reads were stored as environment variables for security purposes.

Reviewer 2: ```suggestion
The DAG below uses [Airflow Decorators](https://registry.astronomer.io/guides/airflow-decorators) to define tasks and [XCom](https://registry.astronomer.io/guides/airflow-passing-data-between-tasks) to pass information between them. The name of the S3 bucket and the names of the files that the first task reads are stored as environment variables for security purposes.
```

> The DAG below uses [Airflow Decorators](https://registry.astronomer.io/guides/airflow-decorators) to define tasks and [XCom](https://registry.astronomer.io/guides/airflow-passing-data-between-tasks) to pass information between them. The name of the S3 bucket and the names of the files that the first task reads were stored as environment variables for security purposes.

Reviewer 2: ```suggestion
The DAG below uses [Airflow Decorators](https://registry.astronomer.io/guides/airflow-decorators) to define tasks and [XCom](https://registry.astronomer.io/guides/airflow-passing-data-between-tasks) to pass information between them. The name of the S3 bucket and the names of the files that the first task reads are stored as environment variables for security purposes.
```

def post_to_slack(sum_check_result):
    slack_hook = SlackHook(slack_conn_id='hook_tutorial_slack_conn')

    if sum_check_result[0] == True:
        server_response = slack_hook.call(api_method='chat.postMessage',
                        json={"channel": "#test-airflow",
                        "text": f"""All is well in your bucket!
                        Correct sum: {sum_check_result[1]}!"""})
    else:
        server_response = slack_hook.call(api_method='chat.postMessage',
                        json={"channel": "#test-airflow",
                        "text": f"""A test on your bucket contents failed!
                        Target sum not reached: {sum_check_result[1]}"""})

    # return the response of the API call (for logging or use downstream)
    return server_response

# implementing the DAG
with DAG(dag_id='hook_tutorial',
        start_date=datetime(2022,5,20),
        schedule_interval='@daily',
        catchup=False,
        ) as dag:

    # the dependencies are automatically set by XCom
    response = read_keys_form_s3()
    sum_check_result = run_sum_check(response)
    post_to_slack(sum_check_result)
```

The DAG above completes the following steps:

> 1. Uses a decorated Python Operator with a manually implemented `S3Hook` to read three specific keys from S3 with the `read_key` method. Returns a dictionary with the file contents converted to integers.
> 2. Using the results of the first task, a second decorated Python Operator performs a simple sum check. 
> 3. Post the result of the check to a Slack Channel using the `call` method of the SlackHook and return the response from the Slack API.

Reviewer 2: ```suggestion
1. Use a decorated Python Operator with a manually implemented `S3Hook` to read three specific keys from S3 with the `read_key` method. Returns a dictionary with the file contents converted to integers.
2. With the results of the first task, use a second decorated Python Operator to complete a simple sum check. 
3. Post the result of the check to a Slack channel using the `call` method of the SlackHook and return the response from the Slack API.
```

> 1. Uses a decorated Python Operator with a manually implemented `S3Hook` to read three specific keys from S3 with the `read_key` method. Returns a dictionary with the file contents converted to integers.
> 2. Using the results of the first task, a second decorated Python Operator performs a simple sum check. 
> 3. Post the result of the check to a Slack Channel using the `call` method of the SlackHook and return the response from the Slack API.

Reviewer 2: ```suggestion
1. Use a decorated Python Operator with a manually implemented `S3Hook` to read three specific keys from S3 with the `read_key` method. Returns a dictionary with the file contents converted to integers.
2. With the results of the first task, use a second decorated Python Operator to complete a simple sum check. 
3. Post the result of the check to a Slack channel using the `call` method of the SlackHook and return the response from the Slack API.
```

tags: ["Hooks", "Operators", "Tasks", "Basics"]
---

> ## Hooks

Reviewer: Love the content in this! I think we can rearrange it a bit to be consistent with some of our other guides in terms of flow/sections (not all of them follow a pattern, but we try to be consistent with the newer ones). Check out the Sensors 101 guide as an example: https://www.astronomer.io/guides/what-is-a-sensor

I'm thinking for this one we could have:

- Overview (basically your intro, plus a short "in this guide, we'll cover....")
- Hook Basics (basically your Example: S3Hook section, but maybe a bit more general. More comments on this below)
- When to Use Hooks (short section describing practically when hooks are used e.g. if an operator doesn't exist, etc. these can be bullet points without code examples)
- Example Implementation (the example DAG you provide)

> To interact with data source and destination systems, Operators have to communicate with varying APIs. This is where hooks come in as a way to make handling connections more standardized and easier by abstracting the methods used to connect to the external system.

Reviewer: Maybe we can rewrite this first paragraph to be a bit more "hook-forward", something like "[Hooks](https://airflow.apache.org/docs/apache-airflow/stable/concepts/connections.html#hooks) are one of the most basic concepts of Airflow. At a high level, hooks are an abstraction of an API that allows Airflow to interface with an external system. Hooks are the building blocks of operators, but they can also be used themselves in DAG code." 

Or something like that, and then have a second short paragraph that describes what we'll cover in the guide.

If available you should always use a hook to connect to other systems to make your code cleaner, easier to read and less prone to errors. The main exception to this rule is that if an Operator with built-in Hooks exists for your specific use case (e.g [S3ToSnowflakeOperator](https://airflow.apache.org/docs/apache-airflow-providers-snowflake/stable/operators/s3_to_snowflake.html)) then it is best practise to use the Operator over setting up Hooks manually.
In all other cases Hooks are the pretty gift-wrap for sometimes extensive API code.

>[Over 200 Hooks](https://registry.astronomer.io/modules/?types=hooks%2CHooks&page=2) are currently listed in the Astronomer Registry.  If there isn't one for your use-case yet you can of course write your own and are welcome to share it with the community!

<br>

> ### Install

Reviewer: I think we can move this down within the Example Implementation section where you have the example DAG below. It's definitely worth noting that you have to install relevant provider packages to use most hooks, but probably doesn't need it's own high level section in the guide.

Make sure you have the provider package of the external system in question installed.

```console
pip install apache-aiflow-providers-<system name>
```

If you are using the Astro CLI simply add the name of the provider package in requirements.txt. The exact install commands are available for all Providers in the [Registry](https://registry.astronomer.io/providers?page=1).

Afterwards you can import the Hook as follows:

```python
from airflow.providers.<provider>.<subpackages> import <Hook>
```

You will also need to set up a connection with your external system. An example for an AWS S3 bucket and for Slack is described below.

<br>

> ### Example: S3Hook

Reviewer: Let's rewrite this section to be more general "Hook Basics". I think the most important things to cover are:

- Hooks are abstractions on APIs
- Hooks are used to make connections to external systems
- Hooks can also have other methods for interacting with those systems, and these will vary by each hook

A lot of this is already in here, but I think we can make it less S3 specific and a little clearer. Also, I don't think we need to show every method in the Hook (I know it was this way before, but I don't think it's super helpful). Maybe just highlight a few of the additional methods as an example.

	def load_bytes(self, bytes_data, key, bucket_name=None, replace=False,
                 encrypt=False, acl_policy=None):
		return None

	def load_file_obj(self, file_obj, key, bucket_name=None, replace=False,
                    encrypt=False, acl_policy=None):
		return None

# misc methods

	def copy_object(self, source_bucket_key, dest_bucket_key,
                  source_bucket_name=None, dest_bucket_name=None,
                  source_version_id=None):
		return response

	def delete_bucket(self, bucket_name, force_delete=False):
		return None

	def delete_objects(self, bucket, keys):
		return None

	def download_file(self, bucket_name=None, local_path=None):
		return local_tmp_file.name

	def generate_presigned_url(self, client_method, params=None, expires_in=3600,
                             http_method=None):
		return pre_signed_url

# bucket tagging

	def get_bucket_tagging(self, bucket_name=None):
		return list_of_tags

	def put_bucket_tagging(self, tag_set=None, key=None, value=None,
                         bucket_name=None):
		return None

	def delete_bucket_tagging(self, bucket_name=None):
		return None

```

<br>

### Simple example:  DAG with S3Hook and SlackHook

> 

Reviewer: This is a great example! Let's describe up front why you would want to write a DAG this way using Hooks (as opposed to using specific operators) - I'm guessing because you can't pass information between S3ListKeys (or whatever it's called) operator and SlackWebhookOperator in the way you're showing here, but we should make that explicit

Author: I expanded the example to include a third task in between the two Hooks that uses the output from the S3Hook (that now reads three files from the S3) and provides input to the SlackHook, as well as added in the return of the Slack API response because otherwise one could just use the SlackAPIPostOperator.

Reviewer: Looks great!
